patches/api/0418-Expose-more-native-map-methods-for-PDC.patch

```diff
@@ -0,0 +1,217 @@
+- #replace(NamespacedKey, PersistentDataType, Object, Object)
+- #replace(NamespacedKey, PersistentDataType, Object)
+
+diff --git a/src/main/java/org/bukkit/persistence/PersistentDataContainer.java b/src/main/java/org/bukkit/persistence/PersistentDataContainer.java
+index d30ad7cae1dfbb47ab1a6e3e2fcd6fb76163fe78..f63d681b7c397c76cba65f859d8fa9162a6596fb 100644
+--- a/src/main/java/org/bukkit/persistence/PersistentDataContainer.java
++++ b/src/main/java/org/bukkit/persistence/PersistentDataContainer.java
+@@ -1,15 +1,22 @@
+ package org.bukkit.persistence;
+ 
+-import java.util.Set;
++import net.kyori.adventure.key.Key;
+ import org.bukkit.NamespacedKey;
+ import org.jetbrains.annotations.NotNull;
+ import org.jetbrains.annotations.Nullable;
+ 
++import java.util.Iterator;
++import java.util.NoSuchElementException;
++import java.util.Set;
++import java.util.function.BiFunction;
++import java.util.function.Function;
++import java.util.function.Supplier;
++
+ /**
+  * This interface represents a map like object, capable of storing custom tags
+  * in it.
+  */
+-public interface PersistentDataContainer {
++public interface PersistentDataContainer extends Iterable<net.kyori.adventure.key.Key> {
+ 
+     /**
+      * Stores a metadata value on the {@link PersistentDataHolder} instance.
+@@ -197,5 +204,171 @@ public interface PersistentDataContainer {
+     default void readFromBytes(byte @NotNull [] bytes) throws java.io.IOException {
+         this.readFromBytes(bytes, true);
+     }
++
++    /**
++     * @return The iterator bound to the set from {@link #getKeys()}
++     */
++    @Override
++    default Iterator<Key> iterator() {
++        return getKeys().iterator();
++    }
++
++    /**
++     * Will set a metadata value on the {@link PersistentDataHolder} instance if the value is not currently set
++     *
++     * @param key   the key this value will be stored under
++     * @param type  the type this tag uses
++     * @param value the value stored in the tag
++     * @param <T>   the generic java type of the tag value
++     * @param <Z>   the generic type of the object to store
++     *
++     * @throws NullPointerException     if the key is null
++     * @throws NullPointerException     if the type is null
++     * @throws NullPointerException     if the value is null. Removing a tag should
++     *                                  be done using {@link #remove(NamespacedKey)}
++     * @throws IllegalArgumentException if no suitable adapter will be found for
++     *                                  the {@link PersistentDataType#getPrimitiveType()}
++     */
++    <T, Z> void setIfAbsent(@NotNull NamespacedKey key, @NotNull PersistentDataType<T, Z> type, @NotNull Z value);
++
++    /**
++     * Attempts to set a metadata value to your {@link PersistentDataHolder} instance if the supplied key isn't bound,
++     * or the current value is null. If your {@param mappingFunction} is null, it will not set the value
```

Comment:
This makes it sound like if the parameter passed is null, not if the return value of the mappingFunction is null.